subsys/testsuite/ztest/src/ztress.c

```diff
@@ -0,0 +1,421 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <ztress.h>
+#include <sys/printk.h>
+#include <random/rand32.h>
+#include <string.h>
+
+
+/* Timer used for adjusting contexts backoff time to get optimal CPU load. */
+static void ctrl_timeout(struct k_timer *timer);
+K_TIMER_DEFINE(ctrl_timer, ctrl_timeout, NULL);
+
+/* Timer used for reporting test progress. */
+static void progress_timeout(struct k_timer *timer);
+K_TIMER_DEFINE(progress_timer, progress_timeout, NULL);
+
+/* Timer used for higher priority context. */
+static void ztress_timeout(struct k_timer *timer);
+K_TIMER_DEFINE(ztress_timer, ztress_timeout, NULL);
+
+/* Timer handling test timeout which ends test prematurely. */
+static k_timeout_t timeout;
+static void test_timeout(struct k_timer *timer);
+K_TIMER_DEFINE(test_timer, test_timeout, NULL);
+
+static atomic_t active_cnt;
+static struct k_thread threads[CONFIG_ZTRESS_MAX_THREADS];
+static k_tid_t tids[CONFIG_ZTRESS_MAX_THREADS];
+
+static uint32_t context_cnt;
+struct ztress_context_data *tmr_data;
+
+static atomic_t active_mask;
+static uint32_t preempt_cnt[CONFIG_ZTRESS_MAX_THREADS];
+static uint32_t exec_cnt[CONFIG_ZTRESS_MAX_THREADS];
+static k_timeout_t backoff[CONFIG_ZTRESS_MAX_THREADS];
+static k_timeout_t init_backoff[CONFIG_ZTRESS_MAX_THREADS];
+K_THREAD_STACK_ARRAY_DEFINE(stacks, CONFIG_ZTRESS_MAX_THREADS, CONFIG_ZTRESS_STACK_SIZE);
+static k_tid_t idle_tid;
+
+#define THREAD_NAME(i, _) STRINGIFY(ztress_##i),
+
+static const char * const thread_names[] = {
+	UTIL_LISTIFY(CONFIG_ZTRESS_MAX_THREADS, THREAD_NAME, _)
+};
+
+struct ztress_runtime {
+	uint32_t cpu_load;
+	uint32_t cpu_load_measurments;
+};
+
+static struct ztress_runtime rt;
+
+static void test_timeout(struct k_timer *timer)
+{
+	ztress_abort();
+}
+
+/* Ratio is 1/16, e.g using ration 14 reduces all timeouts by multipling it by 14/16.
+ * 16 fraction is used to avoid dividing which may take more time on certain platforms.
+ */
+static void adjust_load(uint8_t ratio)
+{
+	for (uint32_t i = 0; i < context_cnt; i++) {
+		uint32_t new_ticks = ratio * (uint32_t)backoff[i].ticks / 16;
+
+		backoff[i].ticks = MAX(4, new_ticks);
+	}
+}
+
+static void progress_timeout(struct k_timer *timer)
+{
+	struct ztress_context_data *thread_data = k_timer_user_data_get(timer);
+	uint32_t progress = 100;
+	uint32_t cnt = context_cnt;
+
+	if (tmr_data) {
+		cnt--;
+		if (tmr_data->exec_cnt && exec_cnt[cnt]) {
```

Comment:
As I understand the current coding guidelines (https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#coding-guidelines), this would be better expressed as ...

`if ((tmr_data->exec_cnt != 0) && (exec_cnt[cnt] != 0)) {`